### PR TITLE
Simpler Python Parameter class

### PR DESCRIPTION
**Pull request by codingS3b**
This PR gets rid of `UiParameter`, `LinearScaledParameter` and `LogScaledParameter` and instead introduces a single `Parameter` class which has the option to pass in arbitrary conversion functions to handle different scales on the UI and on the internal PIC scale of each parameter.
It also makes use of the `pint` package for easier unit conversion.

### CODE REVIEW

File: lib/python/picongpu/input/parameters.py [hunk 1]
```diff
@@ -5,17 +5,41 @@
...
+import pint
+ureg = pint.UnitRegistry()
 
 
 class Parameter(object):
     """
-    Base class for exposable parameters.
+    Parameter that can have a corresponding representation as a jupyter widget
+    within interactive jupyter notebooks.
+    The widget allows the user to adjust values as he likes but this class
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: he/she

File: lib/python/picongpu/input/parameters.py [hunk 2]
```diff
@@ -5,17 +5,41 @@
...
+    precise number is needed (--> log scale conversion)
+    or
+    user adjusts time in nanoseconds but internally the precise number is
+    needed (--> linear conversion)
+    )
+    Also the kind of widget representation should depend on the attributes of
+    this class. Either as MultiRangeSlider or as MultiSelection or as Checkbox
+    depending on the 'values' (discrete) or 'range' (continuous)
+
+    Stepsizes will be handled by the widget representation, not here.q
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: ~~q~~

File: lib/python/picongpu/input/parameters.py [hunk 3]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
-            The value of the parameter object. If it is not provided, the value
-            of the 'default' argument is used.
-        dtype: type (optional)
-            The data type of the parameter value. If it is not provided, the
-            type of the 'default' argument is used.
+            The measurement unit of this parameter as shown on the UI side.
+            Implicitely we assume the pic unit is SI. If this is not possible,
+            the converter needs to translate those scales.
+        default: float, int or string
+            The default value for this parameter (on the UI scale) used when\
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: \ ?

File: lib/python/picongpu/input/parameters.py [hunk 4]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
+        values: list
+            list of discrete options for the parameter as shown on UI side.
+            Only one of the attributes range/value can be given.
+        range: tuple
+            start and stop value for the selectable range on UI side.
+            Only one of the attributes range/value can be given.
+        label: string [optional]
+            Overwrite the name for UI
+        pic_to_SI: callable, e.g. lambda function
+            Specifies how to transform a value on the pic scale to an\
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: \ ?

File: lib/python/picongpu/input/parameters.py [hunk 5]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
+            # raise ValueError("Need either 'values' or 'range' parameter!")
+            self.values = [self.default]
+            print("WARNING: Neither 'values' nor 'range' was given, setting"
+                  " 'values' to ", self.values)
+        if values is not None:
+            if not isinstance(values, collections.Iterable):
+                self.values = [values]
+            if not self.values:
+                self.values = [self.default]
+                print("WARNING: Values attribute can not be an empty\
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: I think
```python
print("WARNING: Values attribute can not be an empty"
      "iterable! Settting values to: ", self.values)
```

should work

File: lib/python/picongpu/input/parameters.py [hunk 6]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
-        Returns
-        -------
-        Whatever the set_value function returns.
         """
+        if self.values is not None:
+            # check for valid values
+            res = all([v in self.values for v in vals])
+            if not res:
+                raise ValueError(
+                    "Invalid values found! Values should be elements of\
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]:
```python
"Invalid values found! Values should be elements of "
"self.values!")

File: lib/python/picongpu/input/parameters.py [hunk 7]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
+                     self.values!")
+        else:
+            # check for valid range
+            res = all([self.range[0] <= v <= self.range[1] for v in vals])
+            if not res:
+                raise ValueError("Invalid values found! Values should be\
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: see above

File: lib/python/picongpu/input/parameters.py [hunk 8]
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
...
-        scale_factor: float
-            The factor for the linear transformation of slider ui value to
-            internal value.
+        A list of converted values.
         """
-        self.scale_factor = scale_factor
-        # since base class UiParameter now calls the set_value() to carry out
-        # linear scaling of the value, we need to define the scale_factor
-        # before entering the constructor since it is used in set_value()
+        # todo check if vals are in specified range
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: open todo :-)

**Comment by codingS3b** [comment]
Thanks for reviewing, I fixed the linebreaks and incorporated the other changes.
Would it be alright to also include the updated params.py for LWFA example in this PR that already uses the new `Parameter` class?

**Comment by ax3l** [comment]
Good point, go ahead and keep it in consistent state.

File: lib/python/picongpu/input/parameters.py [hunk 9]
```diff
@@ -32,8 +32,7 @@ class Parameter(object):
     this class. Either as MultiRangeSlider or as MultiSelection or as Checkbox
     depending on the 'values' (discrete) or 'range' (continuous)
 
-    Stepsizes will be handled by the widget representation, not here.q
-
+    Stepsizes will be handled by the widget representation, not here.
```
Comment (ax3l) [inline lib/python/picongpu/input/parameters.py]: ups, step sizes is two words

Comment (ax3l) [reply lib/python/picongpu/input/parameters.py]: Probably "stepping" is easier to understand?

Comment (codingS3b) [reply lib/python/picongpu/input/parameters.py]: Agreed.

**Pull request merged**